File: TOI neiroseti (python)/02/02.cs
// https://www.youtube.com/watch?v=uzaVPg79bfQ
List<CharCode> dictionary = new();

string str = "testing string";
CheckNode(ToTree(FindProbability(str, FindLetters(str))), "");
var code = Code(str);
var decode = Decode(code);
Console.WriteLine($"Source message - {str}");
Console.WriteLine($"Encode message - {code}");
Console.WriteLine($"Decode message - {decode}");
Print();

char[] FindLetters(string str) {
    return str.Distinct().ToArray();
}

List<CharFrequency> FindProbability(string str, char[] letters) {
    var probability = new List<CharFrequency>();
    foreach (char c in letters) {
        var match = str.Where(x => x == c);
        probability.Add(new() { Used = match.Count(), Char = c });
    }
    probability = probability.OrderBy(e => e.Used).ToList();
    return probability;
}


Node ToTree(List<CharFrequency> list) {
    var nodes = new List<Node>();
    foreach (CharFrequency item in list)
    {
        nodes.Add(new(item.Char, item.Used));
    }
    nodes = nodes.OrderBy(e => e.Order).ToList();
    while (nodes.Count() > 1)
    {
        var left = nodes[0];
        var right = nodes[1];
        nodes.RemoveAt(0);
        nodes.RemoveAt(0);
        nodes.Add(new(left, right));
        nodes = nodes.OrderBy(e => e.Order).ToList();
    }
    return nodes[0];
}


void CheckNode(Node node, string code)
{
    if (node.Left is not null)
    {
        CheckNode(node.Left, code + "0");
    }
    if (node.Right is not null)
    {
        CheckNode(node.Right, code + "1");
    }
    if (node.Char is not null)
    {
        dictionary.Add(new() { Char = (char)node.Char, Code = code });
    }
}

string Code(string str)
{
    string result = "";
    foreach (var s in str)
    {
        result += dictionary.Where(e => e.Char == s).Select(e => e.Code).FirstOrDefault();
    }
    return result;
}


string Decode(string code)
{
    string result = "", temp = "";
    while (code.Length > 0)
    {
        temp += code.Substring(0, 1);
        code = code.Substring(1, code.Length - 1);
        if (dictionary.Any(e => e.Code == temp))
        {
            result += dictionary.Where(e => e.Code == temp).Select(e => e.Char).FirstOrDefault();
            temp = "";
        }
    }
    return result;
}

void Print()
{
    Console.WriteLine("");
    foreach (var el in dictionary)
    {
        Console.WriteLine($"Symbol: {el.Char} Code: {el.Code}");
    }
    Console.WriteLine("");
}


public class Node
{
    public char? Char;
    public int Order;
    public Node Left;
    public Node Right;

    public Node(char ch, int priority)
    {
        Char = ch;
        Order = priority;
    }
    public Node(Node left, Node right)
    {
        Left = left;
        Right = right;
        Order = left.Order + right.Order;
    }
}
record struct CharCode (char Char, string Code);
record struct CharFrequency(char Char, int Used);

File: TOI neiroseti (python)/02/02.py
from __future__ import annotations # объявление ссылок на класс внутри самого класса (Node.Left, Node.Right)

# для словаря (символ - код)
class DictionaryEntry:
    Char: str
    Code: str
    
    def __init__(self, char, code):
        self.Char = char
        self.Code = code

# символ и частота его появления в тексте
class CharFrequency:
    Char: str
    Used: int
    
    def __init__(self, char, used):
        self.Char = char
        self.Used = used

# узел дерева: символ, его вес и соседи слева и справа
class Node:
    Char: str = None
    Order: int
    Left: Node
    Right: Node
    
    def __init__(self, ch = None, priority = None, left = None, right = None):
        if (left == None and right == None):
            self.Char = ch
            self.Order = priority
        else:
            self.Left = left
            self.Right = right
            self.Order = left.Order + right.Order

def TreeBase(str, letters):
# отсортированный список символов и их частот в тексте
    list = []
    for c in letters:
        list.append(CharFrequency(c, str.count(c)))
    list = sorted(list, key=lambda x: x.Used)
# построение дерева и возврат его корня
    nodes = []
    for item in list:
        nodes.append(Node(item.Char, item.Used))
    nodes = sorted(nodes, key=lambda x: x.Order)
    while (len(nodes) > 1):
        left = nodes[0]
        right = nodes[1]
        nodes = nodes[2:]
        nodes.append(Node(None, None, left, right))
        nodes = sorted(nodes, key=lambda x: x.Order)
    return nodes[0]

# заполнение словаря кодами с использованием построенного дерева
def FillDictionary(node, code = ""):
    if (hasattr(node, "Left") and node.Left != None): FillDictionary(node.Left, code + "0")
    if (hasattr(node, "Right") and node.Right != None): FillDictionary(node.Right, code + "1")
    if (hasattr(node, "Char") and node.Char != None):
        ch = DictionaryEntry(node.Char, code)
        dictionary.append(ch)

def HuffmanEncode(str):
    res = ""
    for s in str:
        res += next((x.Code for x in dictionary if x.Char == s), "")
    return res

def HuffmanDecode(code):
    res = ""
    temp = ""
    while (len(code) > 0):
        temp += code[0]
        code = code[1:len(code)]
        if (any([e.Code == temp for e in dictionary])):
            res += next((e.Char for e in dictionary if e.Code == temp), "")
            temp = ""
    return res

dictionary = []
print("Введите любую строку:")
text = input()
FillDictionary(TreeBase(text, set(text)));
code = HuffmanEncode(text);
print("Закодированная строка:", code)
decodedStr = HuffmanDecode(code);
print("Декодированная строка:", decodedStr)
for el in dictionary:
    print("Символ: ", el.Char, " Код: ", el.Code)